Revive the pre 4.02.0 compilers

dra27 opened the pull request:
The old OCaml compilers are of interest to ~a few people~at least me and @herbelin (see #5260)! I recall noticing several years ago that 3.07 and 3.08 had stopped building on anything later than Ubuntu 14.04 (IIRC!), but my lazy solution on the very rare occasions that I've "needed" them is to fire up that old VM. It's been gradually spreading, though as gcc and, more relevantly, the linker have evolved. I can't visit actual care homes at the moment, so I spent a little time with some old compilers instead.

As always, the patches carried here are strictly ones required to build an old OCaml with a new C compiler. For 3.07-3.09, this necessitates back-porting the 3.09 addition of `-fPIC` to ocamlopt to 3.07 and 3.08 and the 3.10.0 adjustment to use it by default to all three of those.

- It turns out OCaml 3 is not affected by the `-fcommon` fix (I wasn't interested enough to check why!) so the patches in #16583 are extended to 4.00.0, 4.00.1 and 4.01.0
- `ocaml-base-compiler.3.11.2` contained an external reference to a patch on the OCamlPRO website which is no longer there. This patch is required for any modern binutils and so I've replaced it with the actual commit from the OCaml repo and extended it to 3.11.0 and 3.11.1
- `ocaml-config.1` has never worked on 3.07... I've simply inserted `ocaml-config.0` to handle the change in version number scheme 😂 Note that the change to `ocaml-config.1` will trigger a switch rebuild in opam 2.1 but because `ocaml-config` is a base package, it won't trigger that in opam 2.0. On that basis, I've amended its constraints, rather than a slightly more weird adjustment to `ocaml-base-compiler.3.07*/opam`.

If we're going to have these packages (and, even as we end up splitting opam-repository, they should still be kept) then we should periodically check that they still build. To that end, I've adjusted 3 packages:
- `ocamlfind.1.0.4` is the last version which supported 3.07, so adding it allows `ocamlfind` to be installed in _any_ ocaml switch
- `pcre.6.2.5` allows a library with C stubs to be installed in 3.08+
- `calendar.1.10` - at least at present - allows a single package at a single version to be installed on _all_ supported versions of OCaml

As usual, while touching `ocaml-base-compiler` packages and others, I've updated metadata.

dra27 commented:
I should add all 23 of these compilers are successfully running in Ubuntu 20.10

camelus commented:
Commit: 470f3eb22724fbd3a66f67b5c06e8fe0020e3292

A pull request by opam-seasoned @dra27.

##### :sunny: All lint checks passed <small>470f3eb22724fbd3a66f67b5c06e8fe0020e3292</small>

* These packages passed lint tests: `calendar.1.10`, `calendar.2.03.1`, `calendar.2.03.2`, `calendar.2.04`, `ocaml-base-compiler.3.07+1`, `ocaml-base-compiler.3.07+2`, `ocaml-base-compiler.3.07`, `ocaml-base-compiler.3.08.0`, `ocaml-base-compiler.3.08.1`, `ocaml-base-compiler.3.08.2`, `ocaml-base-compiler.3.08.3`, `ocaml-base-compiler.3.08.4`, `ocaml-base-compiler.3.09.0`, `ocaml-base-compiler.3.09.1`, `ocaml-base-compiler.3.09.2`, `ocaml-base-compiler.3.09.3`, `ocaml-base-compiler.3.10.0`, `ocaml-base-compiler.3.10.1`, `ocaml-base-compiler.3.10.2`, `ocaml-base-compiler.3.11.0`, `ocaml-base-compiler.3.11.1`, `ocaml-base-compiler.3.11.2`, `ocaml-base-compiler.3.12.0`, `ocaml-base-compiler.3.12.1`, `ocaml-base-compiler.4.00.0`, `ocaml-base-compiler.4.00.1`, `ocaml-base-compiler.4.01.0`, `ocaml-config.0`, `ocaml-config.1`, `ocamlfind.1.0.4`, `ocamlnet.3.2.1`, `ocamlnet.3.5.1`, `ocamlnet.3.6.0`, `ocamlnet.3.6.3`, `ocamlnet.3.6.5`, `ocamlnet.3.7.3`, `ocamlnet.3.7.4-1`, `ocamlnet.3.7.4`, `ocamlnet.3.7.5`, `ocamlnet.3.7.6`, `ocamlnet.3.7.7`, `ocamlnet.4.0.1`, `ocamlnet.4.0.2`, `ocamlnet.4.0.4`, `ocamlnet.4.1.7`, `ocamlnet.4.1.8`, `pcre.6.2.5`


---


##### :sunny: Installability check (+4)

* new installable packages (4): `calendar.1.10` `ocaml-config.0` `ocamlfind.1.0.4` `pcre.6.2.5`


---

##### :sun_behind_small_cloud: 37 ignored non-opam files:

* packages/ocaml-base-compiler/ocaml-base-compiler.3.07+1/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.07+1/files/pr2061.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.07+2/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.07+2/files/pr2061.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.07/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.07/files/pr2061.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.08.0/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.08.1/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.08.2/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.08.3/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.08.4/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.0/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.0/files/ocamlopt-fPIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.0/files/pr4439.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.0/files/pr4867.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.1/files/ocamlopt-fPIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.1/files/pr4439.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.1/files/pr4867.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.2/files/ocamlopt-fPIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.2/files/pr4439.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.2/files/pr4867.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.3/files/ocamlopt-fPIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.3/files/pr4439.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.3/files/pr4867.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.10.0/files/pr4867.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.10.1/files/pr4867.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.10.2/files/pr4867.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.11.0/files/pr5237.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.11.1/files/pr5237.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.11.2/files/pr5237.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.4.00.0/files/fix-gcc10.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.4.00.1/files/fix-gcc10.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.4.01.0/files/fix-gcc10.patch
* packages/ocaml-config/ocaml-config.0/files/gen_ocaml_config.ml.in
* packages/ocaml-config/ocaml-config.0/files/ocaml-config.install
* packages/ocamlfind/ocamlfind.1.0.4/files/no-awk-check.patch
* packages/ocamlfind/ocamlfind.1.0.4/files/ocamlfind.install

dra27 commented:
OK, looking at original opam-repo-ci report for this PR, I added commits to put a lowerbound of calendar 2.0 on all packages which didn't have (for those whose OCaml experience isn't old enough, the major change in Calendar 2.0 was packing the library in `CalendarLib`) and also to constrain `opam-depext` for a few missing packages.

OCaml didn't support ARM64 prior to OCaml 4.02.0, so I've updated the existing ARM64 macOS restriction to apply to all platforms. Despite a little attempt, arm32 and ppc64 are both failing, but I don't like to disable them completely since if one were using an appropriate arm32 CPU and ppc64 Linux installation, I expect they would work.

Constraining the packages has expanded the testing of Camelus and opam-repo-ci beyond the original remit of this PR, so here is Camelus's original pristine report!

Commit: 470f3eb22724fbd3a66f67b5c06e8fe0020e3292

A pull request by opam-seasoned @dra27.

##### :sunny: All lint checks passed <small>470f3eb22724fbd3a66f67b5c06e8fe0020e3292</small>

* These packages passed lint tests: `calendar.1.10`, `calendar.2.03.1`, `calendar.2.03.2`, `calendar.2.04`, `ocaml-base-compiler.3.07+1`, `ocaml-base-compiler.3.07+2`, `ocaml-base-compiler.3.07`, `ocaml-base-compiler.3.08.0`, `ocaml-base-compiler.3.08.1`, `ocaml-base-compiler.3.08.2`, `ocaml-base-compiler.3.08.3`, `ocaml-base-compiler.3.08.4`, `ocaml-base-compiler.3.09.0`, `ocaml-base-compiler.3.09.1`, `ocaml-base-compiler.3.09.2`, `ocaml-base-compiler.3.09.3`, `ocaml-base-compiler.3.10.0`, `ocaml-base-compiler.3.10.1`, `ocaml-base-compiler.3.10.2`, `ocaml-base-compiler.3.11.0`, `ocaml-base-compiler.3.11.1`, `ocaml-base-compiler.3.11.2`, `ocaml-base-compiler.3.12.0`, `ocaml-base-compiler.3.12.1`, `ocaml-base-compiler.4.00.0`, `ocaml-base-compiler.4.00.1`, `ocaml-base-compiler.4.01.0`, `ocaml-config.0`, `ocaml-config.1`, `ocamlfind.1.0.4`, `ocamlnet.3.2.1`, `ocamlnet.3.5.1`, `ocamlnet.3.6.0`, `ocamlnet.3.6.3`, `ocamlnet.3.6.5`, `ocamlnet.3.7.3`, `ocamlnet.3.7.4-1`, `ocamlnet.3.7.4`, `ocamlnet.3.7.5`, `ocamlnet.3.7.6`, `ocamlnet.3.7.7`, `ocamlnet.4.0.1`, `ocamlnet.4.0.2`, `ocamlnet.4.0.4`, `ocamlnet.4.1.7`, `ocamlnet.4.1.8`, `pcre.6.2.5`


---


##### :sunny: Installability check (+4)

* new installable packages (4): `calendar.1.10` `ocaml-config.0` `ocamlfind.1.0.4` `pcre.6.2.5`


---

##### :sun_behind_small_cloud: 37 ignored non-opam files:

* packages/ocaml-base-compiler/ocaml-base-compiler.3.07+1/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.07+1/files/pr2061.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.07+2/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.07+2/files/pr2061.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.07/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.07/files/pr2061.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.08.0/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.08.1/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.08.2/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.08.3/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.08.4/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.0/files/PIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.0/files/ocamlopt-fPIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.0/files/pr4439.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.0/files/pr4867.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.1/files/ocamlopt-fPIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.1/files/pr4439.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.1/files/pr4867.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.2/files/ocamlopt-fPIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.2/files/pr4439.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.2/files/pr4867.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.3/files/ocamlopt-fPIC.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.3/files/pr4439.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.09.3/files/pr4867.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.10.0/files/pr4867.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.10.1/files/pr4867.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.10.2/files/pr4867.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.11.0/files/pr5237.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.11.1/files/pr5237.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.3.11.2/files/pr5237.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.4.00.0/files/fix-gcc10.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.4.00.1/files/fix-gcc10.patch
* packages/ocaml-base-compiler/ocaml-base-compiler.4.01.0/files/fix-gcc10.patch
* packages/ocaml-config/ocaml-config.0/files/gen_ocaml_config.ml.in
* packages/ocaml-config/ocaml-config.0/files/ocaml-config.install
* packages/ocamlfind/ocamlfind.1.0.4/files/no-awk-check.patch
* packages/ocamlfind/ocamlfind.1.0.4/files/ocamlfind.install

kit-ty-kate commented:
I did a bit of digging for ppc64 and arm32:
* ppc64 was added in OCaml 4.03
* StrongARM (an implementation of armv4) was supported since OCaml 2.01
* armv5, 6 and 7 were added in OCaml 4.00

In opam, only armv5, 6 and 7 are referred as `arm32`. Previous versions are not.
So I believe it is safe to restrict `arch != "arm32"` to all ocaml packages < 4.00.0 and `arch != "ppc64"` to all ocaml packages < 4.03.0

avsm commented:
thanks!
